pkg/controller/cachemanager/cachemanager_test.go

```diff
@@ -0,0 +1,455 @@
+package cachemanager
+
+import (
+	"context"
+	"testing"
+
+	configv1alpha1 "github.com/open-policy-agent/gatekeeper/v3/apis/config/v1alpha1"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/controller/config/process"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/fakes"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/readiness"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/syncutil"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/syncutil/aggregator"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/util"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/watch"
+	testclient "github.com/open-policy-agent/gatekeeper/v3/test/clients"
+	"github.com/open-policy-agent/gatekeeper/v3/test/testutils"
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+)
+
+var cfg *rest.Config
+
+func TestMain(m *testing.M) {
+	testutils.StartControlPlane(m, &cfg, 3)
+}
+
+// TestCacheManager_AddObject_RemoveObject tests that we can add/ remove objects in the cache.
+func TestCacheManager_AddObject_RemoveObject(t *testing.T) {
+	cm, _, ctx := makeCacheManagerForTest(t, false, false)
+
+	pod := fakes.Pod(
+		fakes.WithNamespace("test-ns"),
+		fakes.WithName("test-name"),
+	)
+	unstructuredPod, err := runtime.DefaultUnstructuredConverter.ToUnstructured(pod)
+	require.NoError(t, err)
+
+	require.NoError(t, cm.AddObject(ctx, &unstructured.Unstructured{Object: unstructuredPod}))
+
+	// test that pod is cache managed
+	opaClient, ok := cm.opa.(*fakes.FakeOpa)
+	require.True(t, ok)
+	require.True(t, opaClient.HasGVK(pod.GroupVersionKind()))
+
+	// now remove the object and verify it's removed
+	require.NoError(t, cm.RemoveObject(ctx, &unstructured.Unstructured{Object: unstructuredPod}))
+	require.False(t, opaClient.HasGVK(pod.GroupVersionKind()))
+}
+
+// TestCacheManager_processExclusion makes sure that we don't add objects that are process excluded.
+func TestCacheManager_processExclusion(t *testing.T) {
+	cm, _, ctx := makeCacheManagerForTest(t, false, false)
+	processExcluder := process.Get()
+	processExcluder.Add([]configv1alpha1.MatchEntry{
+		{
+			ExcludedNamespaces: []util.Wildcard{"test-ns-excluded"},
+			Processes:          []string{"sync"},
+		},
+	})
+	cm.processExcluder.Replace(processExcluder)
+
+	pod := fakes.Pod(
+		fakes.WithNamespace("test-ns-excluded"),
+		fakes.WithName("test-name"),
+	)
+	unstructuredPod, err := runtime.DefaultUnstructuredConverter.ToUnstructured(pod)
+	require.NoError(t, err)
+	require.NoError(t, cm.AddObject(ctx, &unstructured.Unstructured{Object: unstructuredPod}))
+
+	// test that pod from excluded namespace is not cache managed
+	opaClient, ok := cm.opa.(*fakes.FakeOpa)
+	require.True(t, ok)
+	require.False(t, opaClient.HasGVK(pod.GroupVersionKind()))
```

Comment:
Is there a way to make a more specific assertion?  Like "does this specific object exist"?

Reply:
Like an instance of the gvk? I think this assertion is pretty specific as to whether or not the gvk is in the map.

We can use the `contains` maybe for a specific instance

Reply:
Given that we do elsewhere, I think it's better to check that the exact thing is there.  That way a future extension won't yield an accidental pass.